### PR TITLE
SDKS-2617 Revoke old access token upon force refresh

### PR DESCRIPTION
**Pull request by witrisna**
# JIRA Ticket

https://bugster.forgerock.org/jira/browse/SDKS-2617

# Description

Revoke old access token upon force refresh

### CODE REVIEW

File: forgerock-auth/src/main/java/org/forgerock/android/auth/DefaultTokenManager.java [hunk 1]
```diff
@@ -164,6 +164,10 @@ public void refresh(@NonNull AccessToken accessToken, final FRListener<AccessTok
             Listener.onException(listener, new AuthenticationRequiredException("Refresh Token does not exists."));
             return;
         }
+        //If the access token is not expired yet, revoke it
+        if (!accessToken.isExpired()) {
```
Comment (jeyanthanperiyasamy) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/DefaultTokenManager.java:168]: is this check required ?  what happens when you try to revoke the expired token .  i had a android device in the past which my system time is always wrong due to calendar issues. so will there be a situation like server token is not expired and the client is expired with wrong calendar . its a edge case .

**Review by jeyanthanperiyasamy — COMMENTED** [review]
looks good , just a minor comment

**Review by jeyanthanperiyasamy — DISMISSED** [review]

File: forgerock-auth/src/main/java/org/forgerock/android/auth/DefaultTokenManager.java (hunk 1 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/DefaultTokenManager.java:168]: Server returns the token "Expire In" which is the duration for the token, but not the exact time when the token will expired, client-side keep track of when the token expires, however, if the user change device time after receiving the token, the edge case scenario may happen. 

I put the logic to revoke the access token within the defined threshold, and avoid making additional network calls with the expired token. For e.g, if the access token has 30 min lifetime, and user set 10mins threshold, the access token still valid for 10mins.

File: forgerock-auth/src/main/java/org/forgerock/android/auth/OAuth2Client.java [hunk 2]
```diff
@@ -211,13 +211,14 @@ public void onResponse(@NotNull Call call, @NotNull Response response) {
      * @param accessToken The AccessToken to be revoked
      * @param listener    Listener to listen for revoke event
      */
-    public void revoke(@NonNull AccessToken accessToken, final FRListener<Void> listener) {
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/OAuth2Client.java:214]: Update copyright

File: forgerock-auth/src/main/java/org/forgerock/android/auth/OAuth2Client.java [hunk 3]
```diff
@@ -211,13 +211,14 @@ public void onResponse(@NotNull Call call, @NotNull Response response) {
...
+    public void revoke(@NonNull AccessToken accessToken, boolean useRefreshToken, final FRListener<Void> listener) {
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/OAuth2Client.java:214]: Update javadoc to include new `useRefreshToken` parameter. Does the new parameter introduces a break change?

File: forgerock-auth/src/main/java/org/forgerock/android/auth/SessionManager.java [hunk 4]
```diff
@@ -44,8 +45,17 @@ public SessionManager(TokenManager tokenManager, SingleSignOnManager singleSignO
     }
 
     @WorkerThread
-    public void refresh(AccessToken accessToken, FRListener<AccessToken> listener) throws AuthenticationRequiredException {
-        tokenManager.refresh(accessToken, listener);
+    public void refresh(FRListener<AccessToken> listener) {
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/SessionManager.java:48]: Should we include javadoc here? I see more missing javadocs for other methods in this class

File: forgerock-auth/src/test/java/org/forgerock/android/auth/DefaultTokenManagerTest.kt [hunk 5]
```diff
@@ -266,6 +266,7 @@ class DefaultTokenManagerTest : BaseTest() {
     @Test
     @Throws(Throwable::class)
     fun testTokenRefreshWithThreshold() {
+        server.enqueue(MockResponse().setResponseCode(HttpURLConnection.HTTP_OK))
```
Comment (rodrigoareis) [inline forgerock-auth/src/test/java/org/forgerock/android/auth/DefaultTokenManagerTest.kt:269]: Update copyright

**Review by rodrigoareis — CHANGES_REQUESTED** [review]
Overall changes looks good to me. Minor comments to be reviewed

File: forgerock-auth/src/main/java/org/forgerock/android/auth/SessionManager.java (hunk 4 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/SessionManager.java:53]: Javadoc added.

File: forgerock-auth/src/test/java/org/forgerock/android/auth/DefaultTokenManagerTest.kt (hunk 5 above)
Comment (witrisna) [reply forgerock-auth/src/test/java/org/forgerock/android/auth/DefaultTokenManagerTest.kt:269]: Updated

File: forgerock-auth/src/main/java/org/forgerock/android/auth/TokenManager.java [hunk 6]
```diff
@@ -50,9 +50,8 @@ public interface TokenManager {
      *
      * @param accessToken AccessToken
      * @param listener    Listener to listen for refresh event.
-     * @throws AuthenticationRequiredException When failed to Refresh the {@link AccessToken}
      */
-    void refresh(AccessToken accessToken, FRListener<AccessToken> listener) throws AuthenticationRequiredException;
+    void refresh(AccessToken accessToken, FRListener<AccessToken> listener);
```
[the remark this answers is not in the document]
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/TokenManager.java:52]: Updated

File: forgerock-auth/src/main/java/org/forgerock/android/auth/OAuth2Client.java (hunk 3 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/OAuth2Client.java:215]: Javadoc added

File: forgerock-auth/src/main/java/org/forgerock/android/auth/OAuth2Client.java (hunk 2 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/OAuth2Client.java:214]: Updated

**Review by rodrigoareis — APPROVED** [review]
Changes looks good to me

**Review by spetrov — APPROVED** [review]
Looks good! 👍🏻

**Pull request merged**